index.js

```diff
@@ -106,54 +121,75 @@ function createFunction(parameters, defaultType) {
     return fun;
 }
 
-function evaluateCategoricalFunction(parameters, input) {
-    for (var i = 0; i < parameters.stops.length; i++) {
-        if (input === parameters.stops[i][0]) {
-            return parameters.stops[i][1];
-        }
+function evaluateCategoricalFunction(parameters, input, hashedStops) {
+    var value = hashedStops[input];
+    if (value === undefined) {
+      // If the input is not found, return the first value from the original array by default
+      return parameters.stops[0][1];
     }
-    return parameters.stops[0][1];
+
+    return value;
 }
 
 function evaluateIntervalFunction(parameters, input) {
-    for (var i = 0; i < parameters.stops.length; i++) {
-        if (input < parameters.stops[i][0]) break;
-    }
-    return parameters.stops[Math.max(i - 1, 0)][1];
+    // Edge cases
+    var n = parameters.stops.length;
+    if (n === 1) return parameters.stops[0][1];
+    if (input <= parameters.stops[0][0]) return parameters.stops[0][1];
+    if (input >= parameters.stops[n - 1][0]) return parameters.stops[n - 1][1];
+
+    var index = binarySearchForIndex(parameters.stops, input);
+
+    return parameters.stops[index][1];
 }
 
 function evaluateExponentialFunction(parameters, input) {
     var base = parameters.base !== undefined ? parameters.base : 1;
 
-    var i = 0;
-    while (true) {
-        if (i >= parameters.stops.length) break;
-        else if (input <= parameters.stops[i][0]) break;
-        else i++;
-    }
-
-    if (i === 0) {
-        return parameters.stops[i][1];
+    // Edge cases
+    var n = parameters.stops.length;
+    if (n === 1) return parameters.stops[0][1];
+    if (input <= parameters.stops[0][0]) return parameters.stops[0][1];
+    if (input >= parameters.stops[n - 1][0]) return parameters.stops[n - 1][1];
```

Comment:
@mourner Good point-- it is repetitive. I can't see a clear way right now to clean it up without adding additional conditionals in `evaluateExponentialFunction` to signal that we want to clamp the output and not interpolate if the edge cases are hit. The first thing I tried makes performance notably slower on case where we have few stops. Maybe you can offer a suggestion?

I tried this in `evaluateExponentialFunction`:

    // Let binarySearchForIndex() handle the edge cases
    var index = binarySearchForIndex(parameters.stops, input);

    // But make sure not to run interpolation on edge cases
    if (index >= parameters.stops.length - 1)
      return parameters.stops[parameters.stops.length - 1][1];
    if (input <= parameters.stops[0][0])
      return parameters.stops[0][1];

    return interpolate(
            input,
            base,
            parameters.stops[index][0],
            parameters.stops[index + 1][0],
            parameters.stops[index][1],
            parameters.stops[index + 1][1]
    );

and performance went down to:

      >>> Evaluating exponential function for 1000000 iterations with 10 stops
      Only include values within the domain of stops:
      Time: 182.110ms
      Include values outside the domain of stops:
      Time: 97.764ms

vs about 156ms and 80ms before the change.